tensorflow/core/kernels/training_ops_gpu.cu.cc

```diff
@@ -573,6 +639,32 @@ struct ApplyFtrlV2MultiplyLinearByLr<GPUDevice, T> {
   }
 };
 
+template <typename T, typename Tindex, bool has_l2_shrinkage>
+struct SparseApplyFtrl<GPUDevice, T, Tindex, has_l2_shrinkage> {
+  Status operator()(const GPUDevice& d, typename TTypes<T>::Matrix var,
+                    typename TTypes<T>::Matrix accum,
+                    typename TTypes<T>::Matrix linear,
+                    typename TTypes<T>::ConstScalar lr,
+                    typename TTypes<T>::ConstScalar l1,
+                    typename TTypes<T>::ConstScalar l2,
+                    typename TTypes<T>::ConstScalar l2_shrinkage,
+                    typename TTypes<T>::ConstScalar lr_power,
+                    typename TTypes<T>::ConstMatrix grad,
+                    typename TTypes<Tindex>::ConstVec indices, int64 inner_dim,
+                    bool multiply_linear_by_lr) {
+    const Tindex first_dim_size = var.dimension(0);
+    const Tindex grad_size = grad.size();
+    const Tindex indices_size = indices.size();
+    GpuLaunchConfig config = GetGpuLaunchConfig(grad_size, d);
+    return GpuLaunchKernel(
+        SparseApplyFtrlKernel<T, Tindex, has_l2_shrinkage>, config.block_count,
+        config.thread_per_block, 0, d.stream(), var.data(), accum.data(),
+        linear.data(), lr.data(), l1.data(), l2.data(), l2_shrinkage.data(),
+        lr_power.data(), grad.data(), indices.data(), first_dim_size, grad_size,
+        indices_size, multiply_linear_by_lr);
```

Comment:
Can you please add comments of the form `/*l1=*/` at the right places.  Internally we have a clang check that ensures that the comment match the parameter names which avoids bugs from e.g. swapping `l1.data()` and `l2.data()`.

Reply:
Done (also removed the "cpu only" comment).
(Will the clang check work through the GpuLaunchKernel wrapper function?)